File: Fraction Data Type.cs
/*
 * Copyright 2024 Himanshu Bansal
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific
 * language governing permissions and limitations under the License.
 * ==============================================================================
 */
using System;

public class Fraction : IComparable, IComparable<Fraction>
{
	public int Numerator { get; private set; }
	public int Denominator { get; private set; }

	public Fraction(int numerator, int denominator)
	{
		if (denominator == 0)
		{
			throw new ArgumentException("Denominator cannot be zero.");
		}

		Numerator = numerator;
		Denominator = denominator;
		Simplify();
	}

    // Constructor using double type
    public Fraction(double number)
    {
        // Use TryParse to convert double to fraction
        if (TryParse(number.ToString(System.Globalization.CultureInfo.InvariantCulture), out var result) == 1)
        {
            Numerator = result.Numerator;
            Denominator = result.Denominator;
        }
        else
        {
            throw new ArgumentException("Invalid double value for Fraction.");
        }
    }

    // Constructor using string type
    public Fraction(string number)
    {
        if (TryParse(number, out var result) == 1)
        {
            Numerator = result.Numerator;
            Denominator = result.Denominator;
        }
        else
        {
            throw new ArgumentException("Invalid string value for Fraction.");
        }
    }

	private void Simplify()
	{
		int gcd = GCD(Numerator, Denominator);
		Numerator /= gcd;
		Denominator /= gcd;

		if (Denominator < 0)
		{
			Numerator = -Numerator;
			Denominator = -Denominator;
		}
	}

	private int GCD(int a, int b)
	{
		while (b != 0)
		{
			int temp = b;
			b = a % b;
			a = temp;
		}
		return a;
	}

	public override string ToString()
	{
		return $"{Numerator}/{Denominator}";
	}

	public override bool Equals(object? obj)
	{
		if (obj == null)
		{
			return false;
		}
	
		if (obj is int numerator)
		{
			return Numerator == numerator && Denominator == 1;
		}

		if (obj is Fraction otherFraction)
		{
			return Numerator == otherFraction.Numerator && Denominator == otherFraction.Denominator;
		}
		return false;
	}

	public override int GetHashCode()
	{
		unchecked
		{
			int hash = 17;
			hash = hash * 23 + Numerator.GetHashCode();
			hash = hash * 23 + Denominator.GetHashCode();
			return hash;
		}
	}

	// Addition
	public static Fraction operator +(Fraction a, Fraction b)
	{
		int numerator = a.Numerator * b.Denominator + b.Numerator * a.Denominator;
		int denominator = a.Denominator * b.Denominator;
		return new Fraction(numerator, denominator);
	}

	// Subtraction
	public static Fraction operator -(Fraction a, Fraction b)
	{
		int numerator = a.Numerator * b.Denominator - b.Numerator * a.Denominator;
		int denominator = a.Denominator * b.Denominator;
		return new Fraction(numerator, denominator);
	}

	// Multiplication
	public static Fraction operator *(Fraction a, Fraction b)
	{
		int numerator = a.Numerator * b.Numerator;
		int denominator = a.Denominator * b.Denominator;
		return new Fraction(numerator, denominator);
	}

	public static Fraction operator *(int a, Fraction b)
	{
		int numerator = a * b.Numerator;
		int denominator = b.Denominator;
		return new Fraction(numerator, denominator);
	}

	// Division
	public static Fraction operator /(Fraction a, Fraction b)
	{
		if (b.Numerator == 0)
		{
			throw new DivideByZeroException("Cannot divide by zero.");
		}

		int numerator = a.Numerator * b.Denominator;
		int denominator = a.Denominator * b.Numerator;
		return new Fraction(numerator, denominator);
	}


	//comparison
	public int CompareTo(object? obj)
	{
		if (obj == null) return 1;

		if (obj is Fraction otherFraction)
		{
			return CompareTo(otherFraction);
		}
		else if (obj is int intValue)
		{
			return CompareTo(new Fraction(intValue, 1)); // Convert int to Fraction for comparison
		}
		else
		{
			throw new ArgumentException("Object is not a Fraction or int.");
		}
	}

	// Implement IComparable<Fraction> interface
	public int CompareTo(Fraction? other)
	{
		if (other == null) return 1;

		// Compare fractions based on their double values
		double thisValue = (double)Numerator / Denominator;
		double otherValue = (double)other.Numerator / other.Denominator;

		return thisValue.CompareTo(otherValue);
	}

	public static bool operator ==(Fraction a, int b)
	{
		return a.CompareTo(b) == 0;
	}

	public static bool operator !=(Fraction a, int b)
	{
		return a.CompareTo(b) != 0;
	}

	public static bool operator >(Fraction a, int b)
	{
		return a.CompareTo(b) > 0;
	}

	public static bool operator <(Fraction a, int b)
	{
		return a.CompareTo(b) < 0;
	}

	public static bool operator >=(Fraction a, int b)
	{
		return a.CompareTo(b) >= 0;
	}

	public static bool operator <=(Fraction a, int b)
	{
		return a.CompareTo(b) <= 0;
	}

	public static (long numerator, long denominator) DecimalToFraction(double value, double tolerance = 1.0E-10)
	{
		if (double.IsNaN(value) || double.IsInfinity(value))
			throw new ArgumentException("Value must be a finite number.");

		long numerator = 1, denominator = 0;
		long prevNumerator = 0, prevDenominator = 1;
		double remainder = value;

		while (true)
		{
			long integralPart = (long)Math.Floor(remainder);
			long tempNumerator = numerator;
			long tempDenominator = denominator;

			numerator = integralPart * numerator + prevNumerator;
			denominator = integralPart * denominator + prevDenominator;

			prevNumerator = tempNumerator;
			prevDenominator = tempDenominator;

			double approx = (double)numerator / denominator;
			if (Math.Abs(value - approx) < tolerance)
				break;

			remainder = 1.0 / (remainder - integralPart);
		}

		return (numerator, denominator);
	}

	public static int TryParse(string input, out Fraction result)
	{
		result = new Fraction(0, 1);
		if (string.IsNullOrWhiteSpace(input))
		{
			return 0;
		}

		input = input.Replace("\\", "/");
		int count = input.Count(c => c == '/');
		if (count > 1)
		{
			return 0;
		}
		if (input[^1] == '/' || input[0] == '/')
		{
			return 0;
		}
		string[] parts = input.Split('/');
		if (parts.Length == 1)
		{
			double numerator1;
			double denominator1 = 1;
			bool success = double.TryParse(parts[0], System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out numerator1);
			if (success)
			{
				// Convert decimal to fraction
				int decimals = 0;
				string[] splitNum = parts[0].Split('.');
				if (splitNum.Length == 2)
				{
					decimals = splitNum[1].Length;
				}
				if (decimals > 3)
				{
					double tolerance = Math.Pow(10, -decimals);
					(numerator1, denominator1) = DecimalToFraction(numerator1, tolerance);
				}
				else
				{
					(numerator1, denominator1) = DecimalToFraction(numerator1);
				}
				if (Math.Abs(numerator1) > 2147483640 || Math.Abs(denominator1) > 2147483640)
				{
					return -1;
				}
				int numerator = (int)Math.Round(numerator1);
				int denominator = (int)Math.Round(denominator1);
				if (denominator == 0)
				{
					return -2;
				}
				result = new Fraction(numerator, denominator);
				return 1;
			}
			return 0;
		}
		else if (parts.Length == 2)
		{
			int numerator, denominator;
			if (!int.TryParse(parts[0], out numerator) || !int.TryParse(parts[1], out denominator))
			{
				double numerator1, denominator1;
				if (!double.TryParse(parts[0], System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out numerator1) ||
					!double.TryParse(parts[1], System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out denominator1))
				{
					return 0;
				}
				while (Math.Abs(numerator1 % 10) < double.Epsilon && Math.Abs(denominator1 % 10) < double.Epsilon && numerator1 != 0)
				{
					numerator1 /= 10;
					denominator1 /= 10;
				}
				if (Math.Abs(numerator1) > 2147483640 || Math.Abs(denominator1) > 2147483640)
				{
					return -1;
				}
				numerator = (int)Math.Round(numerator1);
				denominator = (int)Math.Round(denominator1);
				if (denominator == 0)
				{
					return -2;
				}
				result = new Fraction(numerator, denominator);
				return 1;
			}
			else if (denominator == 0)
			{
				return -2;
			}
			result = new Fraction(numerator, denominator);
			return 1;
		}
		return 0;
	}
}

File: Fraction_Data_Type.cpp
/*
 * Copyright 2024-2025 Himanshu Bansal
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific
 * language governing permissions and limitations under the License.
 * ==============================================================================
 */
#include <iostream>
#include <string>
#include <cmath>
#include <limits>
#include <cmath>

#include "Essentials.h"
#include "Fraction_Data_Type.h"
#include <vector>

Fraction::Fraction(int numerator, int denominator)
{
	null = false;
	if (denominator == 0)
	{
		throw std::domain_error("Denominator cannot be zero.");
	}

	Numerator = numerator;
	Denominator = denominator;
	Simplify();
}

Fraction::Fraction(double number)
{
	if (!Fraction::TryParse(std::to_string(number), *this)) {
		throw std::invalid_argument("Error happened");
	}
}

Fraction::Fraction(std::string number)
{
	if (!Fraction::TryParse(number, *this)) {
		throw std::invalid_argument("Error happened");
	}
}

void Fraction::Simplify()
{
    if (Denominator == 0) {
        throw std::domain_error("Denominator cannot be zero in Simplify.");
    }
    int gcd = GCD(Numerator, Denominator);
    Numerator /= gcd;
    Denominator /= gcd;

    if (Denominator < 0)
    {
        Numerator = -Numerator;
        Denominator = -Denominator;
    }
}

int Fraction::GCD(int a, int b)
{
	while (b != 0)
	{
		int temp = b;
		b = a % b;
		a = temp;
	}
	return a;
}

std::string Fraction::ToString()
{
	return std::to_string(Numerator) + "/" + std::to_string(Denominator);
}

// Addition
Fraction Fraction::operator+ (const Fraction& b) const
{
	int numerator = Numerator * b.Denominator + b.Numerator * Denominator;
	int denominator = Denominator * b.Denominator;
	return Fraction(numerator, denominator);
}

// Subtraction
Fraction Fraction::operator- (const Fraction& b) const
{
	int numerator = Numerator * b.Denominator - b.Numerator * Denominator;
	int denominator = Denominator * b.Denominator;
	return Fraction(numerator, denominator);
}

// Multiplication
Fraction Fraction::operator* (const Fraction& b) const
{
	int numerator = Numerator * b.Numerator;
	int denominator = Denominator * b.Denominator;
	return Fraction(numerator, denominator);
}

Fraction Fraction::operator* (const int& b) const
{
	int numerator = Numerator * b;
	int denominator = Denominator;
	return Fraction(numerator, denominator);
}

Fraction operator*(int b, const Fraction& frac) {
	int numerator = frac.Numerator * b;
	int denominator = frac.Denominator;
	return Fraction(numerator, denominator);
}
// Division
Fraction Fraction::operator/ (const Fraction& b) const
{
	if (b.Numerator == 0)
	{
		throw std::domain_error("Cannot divide by zero.");
	}
	int numerator = Numerator * b.Denominator;
	int denominator = Denominator * b.Numerator;
	return Fraction(numerator, denominator);
}


//comparison
int Fraction::CompareTo(const int& obj) const
{
	return CompareTo(Fraction(obj, 1)); // Convert int to Fraction for comparison
}

int Fraction::CompareTo(const Fraction& other) const
{
	if (other.null == true) return 1;

	// Compare fractions based on their double values
	double thisValue = (double)Numerator / Denominator;
	double otherValue = (double)other.Numerator / other.Denominator;

	if (thisValue < otherValue) return -1;
	if (thisValue > otherValue) return 1;
	return 0;
}

bool Fraction::operator== (const Fraction& a) const
{
	return CompareTo(a) == 0;
}

bool Fraction::operator== (const int& a) const
{
	return CompareTo(a) == 0;
}

bool Fraction::operator!= (const Fraction& a) const
{
	return CompareTo(a) != 0;
}

bool Fraction::operator!= (const int& a) const
{
	return CompareTo(a) != 0;
}

bool Fraction::operator> (const Fraction& a) const
{
	return CompareTo(a) > 0;
}

bool Fraction::operator> (const int& a) const
{
	return CompareTo(a) > 0;
}

bool Fraction::operator< (const Fraction& a) const
{
	return CompareTo(a) < 0;
}

bool Fraction::operator< (const int& a) const
{
	return CompareTo(a) < 0;
}

bool Fraction::operator>= (const Fraction& a) const
{
	return CompareTo(a) >= 0;
}

bool Fraction::operator>= (const int& a) const
{
	return CompareTo(a) >= 0;
}

bool Fraction::operator<= (const Fraction& a) const
{
	return CompareTo(a) <= 0;
}

bool Fraction::operator<= (const int& a) const
{
	return CompareTo(a) <= 0;
}

std::tuple<long, long> Fraction::DecimalToFraction(double value, double tolerance = 1.0E-10) {
	if (std::isnan(value) || std::isinf(value)) {
        throw std::invalid_argument("Value must be a finite number.");
    }

	long numerator = 1, denominator = 0;
    long prevNumerator = 0, prevDenominator = 1;
    double remainder = value;

    while (true)
    {
        long integralPart = (long)std::floor(remainder);
        long tempNumerator = numerator;
        long tempDenominator = denominator;

        numerator = integralPart * numerator + prevNumerator;
        denominator = integralPart * denominator + prevDenominator;

        prevNumerator = tempNumerator;
        prevDenominator = tempDenominator;

        double approx = (double)numerator / denominator;
        if (std::abs(value - approx) < tolerance)
            break;

        remainder = 1.0 / (remainder - integralPart);
    }

    return std::tuple<long, long>(numerator, denominator);
}

int Fraction::TryParse(std::string input, Fraction& result)
{
    result.null = true;

    if (input.empty() || replaceWord(input, " ", "") == "")
    {
        return 0;
    }

    input = replaceWord(input, "\\", "/");

    int count = 0;
    for (char ch : input) {
        if (ch == '/') {
            count++;
        }
    }

    if (count > 1) {
        return 0;
    }

    if (input[0] == '/' || input[input.size() - 1] == '/') {
        return 0;
    }

    std::vector<std::string> parts = split(input, "/");

    int numerator = 0;
    int denominator = 0;
    if (parts.size() == 1)
    {
        double numerator1 = 1;
        double denominator1 = 1;
        bool success = tryParseDouble(parts[0], numerator1);

        if (success) {
            int decimals = 0;
			std::vector<std::string> decimal_parts = split(parts[0], ".");
            if (decimal_parts.size() == 2) {
                decimals = decimal_parts[1].length();
			}
            if (decimals > 3) {
				double tolerance = std::pow(10, -decimals);
                std::tie(numerator1, denominator1) = DecimalToFraction(numerator1, tolerance);
            }
            else {
                std::tie(numerator1, denominator1) = DecimalToFraction(numerator1);
            }
            if (std::abs(numerator1) > std::numeric_limits<int>::max() || std::abs(denominator1) > std::numeric_limits<int>::max())
            {
                return -1;
            }
            result.null = false;
            int numerator = (int)std::round(numerator1);
            int denominator = (int)std::round(denominator1);
            if (denominator == 0)
            {
                return -2;
            }
            result = Fraction(numerator, denominator);
            return 1;
        }
        return 0;
    }
    else if (parts.size() == 2)
    {
        if (!tryParseInt(parts[0], numerator) || !tryParseInt(parts[1], denominator)) {
            double numerator1 = 1;
            double denominator1 = 1;
            if (!tryParseDouble(parts[0], numerator1) || !tryParseDouble(parts[1], denominator1))
            {
                return 0;
            }

            while (std::round(numerator1 / 10) == numerator1 / 10 && std::round(denominator1 / 10) == denominator1 / 10) {
                numerator1 /= 10;
                denominator1 /= 10;
            }

			if (numerator1 > std::numeric_limits<int>::max() || denominator1 > std::numeric_limits<int>::max())
            {
                return -1;
            }

            numerator = (int)numerator1;
            denominator = (int)denominator1;

            result.null = false;
            if (denominator == 0)
            {
                return -2;
            }

            result.Numerator = numerator;
            result.Denominator = denominator;
            result.Simplify();
            return 1;
        }
        else if (denominator == 0) {
            return -2;
        }
        result.Numerator = numerator;
        result.Denominator = denominator;
        result.Simplify();
        return 1;
    }

    result.null = false;
    if (denominator == 0)
    {
        return -2;
    }

    result.Numerator = numerator;
    result.Denominator = denominator;
    result.Simplify();
    return 1;
}